src/components/form/templates/array-field-collapsible-item.ts

```diff
@@ -70,6 +70,12 @@ export class CollapsibleItemTemplate extends React.Component {
                 className: 'limel-form-array-item--object',
                 ref: 'section',
                 'is-open': this.isOpen,
+                'state-key': encodeURIComponent(
+                    schema?.lime?.stateKey ||
+                        `${schema.items?.$ref}-${this.props.index}-${
+                            data.title || data.property
+                        }`
```

Comment:
Discussion continued from https://github.com/Lundalogik/lime-elements/pull/1431#discussion_r747617427

Reply:
@adrianschmidt wrote:

> > Does this mean that changing the data in the form will create a new key? Will that not break stuff? 😅
> 
> I guess it depends. If you put ever-changing information in it, then, yes, you'd have to give it a key instead. But that's what the `stateKey` property is there for.
> 
> This case (`array-field-collapsible-item.ts`) is the one used in Lime Admin, and the data we use is as follows:
> 
> ![image](https://user-images.githubusercontent.com/154725/142626239-6bf271fd-2923-49d5-a5e2-77b3af257634.png)
> 
> * `schema.items.$ref` is `"#/definitions/CompanyCardSectionSchema"` - this is a reference to stuff in the schema, but it happens to nicely identify the parent section we're in.
> * `this.props.index` is `0` - this is the index of the item we're about to render (the collapsible), among its siblings. So the _Addresses_ section has `1` here.
> * `data.title` is `"untitled"` - The title for the section.
> 
> So the key for this section will be `'#/definitions/CompanyCardSectionSchema - 0 - untitled'`. This means that it won't collide with the section called `untitled`, with index `0` under _Deal Card_, because `$ref` will be different. It also won't remember its state if moved or renamed, and that's intentional. If the section is moved or renamed, for this purpose, we don't consider it to be the same section anymore. If the consumer knows it is, it has to explicitly tell us that by supplying its own `stateKey`.
> 
> If we look at the next level of collapsibles used in Lime Admin, it's the ones for properties, found inside the collapsible we just looked at. Let's look at the _Telephone_ field:
> 
> ![image](https://user-images.githubusercontent.com/154725/142632220-58abd344-6145-470d-b362-c15909c30a7b.png)
> 
> * `schema.items.$ref` is now `'#/definitions/CompanyEditPropertySchema'`
> * `this.props.index` is `1`
> * `data.title` is undefined, so instead, we use
> * `data.property`, which is `phone`
> 
> So the key here is `'#/definitions/CompanyEditPropertySchema - 1 - phone'`.
> 
> In the case of `object-field.ts`, that's the one that's used on the object card in the web client. (I never really did figure out why the same code isn't used in both places. If you know or understand that, I'd love to get up to speed 🙂) The data available here was slightly different, but very similar:
> 
> ![image](https://user-images.githubusercontent.com/154725/142634887-e3b83bbe-d52c-4750-b4d4-33eac0371cf7.png)
> 
> * `props.formContext.schema.$ref` is `'#/definitions/CompanySchema'` - again, this lets us differentiate between sections that have the same name and are in the same place, but on different limetypes.
> * `props.idSchema.$id` is `'root_section1'` - I guess the `root` part might change if we had nestled sections? Not sure. Anyway, the `1` works as the index, and lets us differentiate between two identically named sections on the same limetype.
> * `props.title` is of course `'Addresses'`
> 
> While I mention "limetypes", Lime Elements obviously doesn't know what those are, but also doesn't have to. All that matters is that the string is different between "different places". The question is, can we assume that they will be, at least under "normal circumstances"?
> 
> If not, I guess we'll have to skip this part, and instead add something along these lines in `lime-crm-components` and in the code for Lime Admin (if we want the functionality there too).

Reply:
@jgroth wrote:

> There is no guarantee that `schema.items.$ref` is unique. It probably is in lime-admin because of how we generate the schemas though, but we had to write custom code in the schema in order to get those unique. In a more general case it is more likely that the same sub-schema is used in several places and then `schema.items.$ref` will be the same for those.
> 
> I don't think we should use any data from the form at all (like the title) to generate the key, since that will create a new key whenever the data is changed. The way we use that at the moment to generate the section headers is already too hackish and I would prefer if we don't add any more to that logic 😄
> 
> Would it not be enough to use the property path from the root to the section to uniquely identify it? E.g. `card.sections[1].fields[1].layout`

> I guess we also need an ID for the whole form, so the property path might not be enough 🤔

Reply:
This was always the shakiest part, and I'm not surprised if it doesn't hold up. I think we probably need to just let the consumer identify the different collapsibles, and by then it's probably best to just let the consumer keep track of the state too. We already provide the possibility to set the state as open or closed from the outside, via the `isOpen` property, and it's starting to look more and more likely to me that that's the way to go.